app/models/piece.rb

```diff
@@ -1,8 +1,105 @@
-class Piece < ApplicationRecord
-  belongs_to :user
-  belongs_to :game
-
-  def self.types
-    %w(King Queen Rook Bishop Knight Pawn)
-  end
-end
+class Piece < ApplicationRecord
+  belongs_to :user
+  belongs_to :game
+
+  def self.types
+    %w[King Queen Rook Bishop Knight Pawn]
+  end
+
+  def obstructed?(destination_x, destination_y)
+    @destination_x = destination_x
+    @destination_y = destination_y
+
+    destination_tests = [horizontal?, vertically?, diagonal_up_right_down_left?, diagonal_up_left_down_right?]
+
+    if destination_tests.include?(true)
+      return true
+    elsif invalid_input?(destination_tests)
+      return 'Error: Invalid Input'
+    else
+      false
+    end
+  end
+
+  def invalid_input?(destination_tests)
+    invalid_input = @destination_y > 7 || @destination_x > 7 || @destination_y < 0 || @destination_x < 0
+    return true if (!destination_tests.include?(true) && !destination_tests.include?(false)) || invalid_input
+  end
+
+
+  def horizontal?

```

Comment:
Overall this looks really good. I like how you separated it out into multiple methods.

Some ways to improve it.

1. I always recommend avoiding instance (`@var`) variables inside of models. It makes it hard to keep track of "state" in your head when reading code. What you can do instead is for each method, pass the variables needed to the method.

So for example, your `horizontal?` method. Would become this `horizontal?(destination_x, destination_y)`.

This makes the method easier to understand, because a certain input always results in a certain output.

2. For naming, try to make what each method does really obvious. This is important when working on a codebase with a team. Instead of `horizontal?`. You could do `horizontal_obstruction?`.